### PR TITLE
Fix nil pointer exception when deleting MachinePools

### LINKED ISSUE #4012: Removing a MachinePool resources lead capz-controller-manager to crash loop
/kind bug

**What steps did you take and what happened:**
- Create a cluster with an AzureMachinePool (vmss)
- Removing the MachinePool ressource defining the MachinePool will make capz manager crashing in loop with the the following go dump:

```
│ capz-controller-manager-5f674d4bcd-q2vgn I0919 19:52:34.911596       1 azuremachine_controller.go:243] controllers.AzureMachineReconciler.reconcileNormal "msg"="Reconciling AzureMachine" "AzureMachine"={"name":"test-control-plane-4pv28","namespace":"default"} "controller"="azuremachine" "controllerGroup"="infrastructure.cluster.x-k8s.io" "co │
│ ntrollerKind"="AzureMachine" "name"="test-control-plane-4pv28" "namespace"="default" "reconcileID"="c04c7b25-08d8-45f3-be91-bc49eac2a2c2" "x-ms-correlation-request-id"="578c8ff8-e7cc-4176-a5d4-3dcfcced67e1"                                                                                                                                          │
│ capz-controller-manager-5f674d4bcd-q2vgn I0919 19:52:34.912608       1 controller.go:114]  "msg"="Observed a panic in reconciler: runtime error: invalid memory address or nil pointer dereference" "AzureMachinePool"={"name":"test-mp-0","namespace":"default"} "controller"="azuremachinepool" "controllerGroup"="infrastructure.cluster.x-k8s.io" " │
│ controllerKind"="AzureMachinePool" "name"="test-mp-0" "namespace"="default" "reconcileID"="16729707-5076-4a09-a7c0-ca1afbbe62de"                                                                                                                                                                                                                        │
│ capz-controller-manager-5f674d4bcd-q2vgn panic: runtime error: invalid memory address or nil pointer dereference [recovered]                                                                                                                                                                                                                            │
│ capz-controller-manager-5f674d4bcd-q2vgn     panic: runtime error: invalid memory address or nil pointer dereference                                                                                                                                                                                                                                    │
│ capz-controller-manager-5f674d4bcd-q2vgn [signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x1d777bc]                                                                                                                                                                                                                                       │
│ capz-controller-manager-5f674d4bcd-q2vgn                                                                                                                                                                                                                                                                                                                │
│ capz-controller-manager-5f674d4bcd-q2vgn goroutine 536 [running]:                                                                                                                                                                                                                                                                                       │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile.func1()                                                                                                                                                                                                                         │
│ capz-controller-manager-5f674d4bcd-q2vgn     /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:115 +0x1fa                                                                                                                                                                                                        │
│ capz-controller-manager-5f674d4bcd-q2vgn panic({0x21ffde0, 0x3cacae0})                                                                                                                                                                                                                                                                                  │
│ capz-controller-manager-5f674d4bcd-q2vgn     /usr/local/go/src/runtime/panic.go:884 +0x213                                                                                                                                                                                                                                                              │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/cluster-api-provider-azure/azure/scope.(*MachinePoolScope).ScaleSetSpec(0xc000d4af00, {0x29b0a20?, 0xc000ae07e0?})                                                                                                                                                                                 │
│ capz-controller-manager-5f674d4bcd-q2vgn     /workspace/azure/scope/machinepool.go:182 +0xdc                                                                                                                                                                                                                                                            │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/cluster-api-provider-azure/azure/services/scalesets.(*Service).Delete(0xc0007a3840, {0x29b0a58?, 0xc000ae0600?})                                                                                                                                                                                   │
│ capz-controller-manager-5f674d4bcd-q2vgn     /workspace/azure/services/scalesets/scalesets.go:148 +0xde                                                                                                                                                                                                                                                 │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/cluster-api-provider-azure/exp/controllers.(*azureMachinePoolService).Delete(0xc000ae0570, {0x29b0a58?, 0xc000f46fc0?})                                                                                                                                                                            │
│ capz-controller-manager-5f674d4bcd-q2vgn     /workspace/exp/controllers/azuremachinepool_reconciler.go:109 +0x102                                                                                                                                                                                                                                       │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/cluster-api-provider-azure/exp/controllers.(*AzureMachinePoolReconciler).reconcileDelete(0xc000050870, {0x29b0a20?, 0xc00003c6f0?}, 0xc000d4af00, 0xc000ba8a80)                                                                                                                                    │
│ capz-controller-manager-5f674d4bcd-q2vgn     /workspace/exp/controllers/azuremachinepool_controller.go:390 +0x198                                                                                                                                                                                                                                       │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/cluster-api-provider-azure/exp/controllers.(*AzureMachinePoolReconciler).Reconcile(0xc000050870, {0x29b0a58?, 0xc00003c4e0?}, {{{0xc0009d8989?, 0x7?}, {0xc0009d8980?, 0x9?}}})                                                                                                                    │
│ capz-controller-manager-5f674d4bcd-q2vgn     /workspace/exp/controllers/azuremachinepool_controller.go:260 +0xcd6                                                                                                                                                                                                                                       │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/cluster-api-provider-azure/pkg/coalescing.(*reconciler).Reconcile(0xc00067f640, {0x29b0a58?, 0xc000705f80?}, {{{0xc0009d8989?, 0x0?}, {0xc0009d8980?, 0x40e1c7?}}})                                                                                                                                │
│ capz-controller-manager-5f674d4bcd-q2vgn     /workspace/pkg/coalescing/reconciler.go:109 +0x3eb                                                                                                                                                                                                                                                         │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Reconcile(0x29b0a58?, {0x29b0a58?, 0xc000705f80?}, {{{0xc0009d8989?, 0x20e29a0?}, {0xc0009d8980?, 0x29924e8?}}})                                                                                                                          │
│ capz-controller-manager-5f674d4bcd-q2vgn     /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:118 +0xc8                                                                                                                                                                                                         │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler(0xc0006e7220, {0x29b09b0, 0xc0006dddb0}, {0x237c700?, 0xc000236620?})                                                                                                                                                    │
│ capz-controller-manager-5f674d4bcd-q2vgn     /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:314 +0x377                                                                                                                                                                                                        │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem(0xc0006e7220, {0x29b09b0, 0xc0006dddb0})                                                                                                                                                                              │
│ capz-controller-manager-5f674d4bcd-q2vgn     /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:265 +0x1d9                                                                                                                                                                                                        │
│ capz-controller-manager-5f674d4bcd-q2vgn sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2.2()                                                                                                                                                                                                                           │
│ capz-controller-manager-5f674d4bcd-q2vgn     /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:226 +0x85                                                                                                                                                                                                         │
│ capz-controller-manager-5f674d4bcd-q2vgn created by sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).Start.func2                                                                                                                                                                                                                    │
│ capz-controller-manager-5f674d4bcd-q2vgn     /go/pkg/mod/sigs.k8s.io/controller-runtime@v0.15.1/pkg/internal/controller/controller.go:222 +0x587 
 ```

**What did you expect to happen:**
The MachinePool and AzureMachinePool  resources should be removed after the vmss being removed from the deployed infrastructure.

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- cluster-api-provider-azure version: 1.11.0 (tested in 1.10.4, no issue with it)
- Kubernetes version: (use `kubectl version`):  1.27.3 (kind)
- OS (e.g. from `/etc/os-release`): Ubuntu 22.04

### PR DESCRIPTION
**Pull request by Jont828**
**What type of PR is this?**
/kind bug

**What this PR does / why we need it**: We expect the MachinePool cache to be initialized before getting the ScaleSetSpec. However, we only initialize it in reconcileNormal in azuremachinepool.go, so we have a panic trying to get the ScaleSetSpec for deletion. This fix only sets the fields using cache values if the cache is not nil to allow the spec to be available for deletion.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #4012

**Special notes for your reviewer**:

- [ ] cherry-pick candidate

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Review by CecileRobertMichon — COMMENTED** [review]
Good find! overall lgtm. Looks like there are some leftover debug print/log statements, can you please clean those up for merge?

**Comment by Jont828** [comment]
Sure thing, just wanted to make sure the tests passed before removing logs!

**Review by nojnhuh — COMMENTED** [review]
lgtm aside from one comment.

/cherry-pick release-1.11

File: azure/scope/machinepool.go [hunk 1]
```diff
@@ -209,16 +203,27 @@ func (m *MachinePoolScope) ScaleSetSpec(ctx context.Context) azure.ResourceSpecG
...
-		SKU:                          m.cache.VMSKU,
-		VMImage:                      m.cache.VMImage,
-		BootstrapData:                m.cache.BootstrapData,
-		ShouldPatchCustomData:        shouldPatchCustomData,
-		MaxSurge:                     m.cache.MaxSurge,
 	}
+
+	if m.cache != nil {
+		if m.HasReplicasExternallyManaged(ctx) {
+			log.Info("machinepool cache is", "cache", m.cache)
```
Comment (CecileRobertMichon) [inline azure/scope/machinepool.go:213]: this looks like a leftover log statement

**Review by mboersma — COMMENTED** [review]
Agree, this looks good but for the added log lines.

**Review by nojnhuh — COMMENTED** [review]
/lgtm
/approve

**Pull request merged**